crates/sc-proof-of-time/src/verifier.rs

```diff
@@ -0,0 +1,265 @@
+    ///
+    /// NOTE: Potentially much slower than checkpoints, prefer [`Self::verify_checkpoints()`]
+    /// whenever possible.
+    pub async fn is_proof_valid(
+        &self,
+        mut seed: PotSeed,
+        slot_iterations: NonZeroU32,
+        slots: Slot,
+        proof: PotProof,
+    ) -> bool {
+        let mut slots = u64::from(slots);
+
+        loop {
+            if slots == 0 {
+                return proof.seed() == seed;
+            }
+
+            // TODO: This "proxy" is a workaround for https://github.com/rust-lang/rust/issues/57478
+            let (result_sender, result_receiver) = oneshot::channel();
+            std::thread::spawn({
+                let verifier = self.clone();
+
+                move || {
+                    futures::executor::block_on({
+                        async move {
+                            // Result doesn't matter here
+                            let _ = result_sender
+                                .send(verifier.derive_next_seed(seed, slot_iterations).await);
+                        }
+                    });
+                }
+            });
+
+            seed = match result_receiver.await {
+                Ok(Some(seed)) => seed,
+                _ => {
+                    return false;
+                }
+            };
+
+            slots -= 1;
+        }
+    }
+
+    /// Derive next seed, proving might be used if necessary
+    // TODO: False-positive, lock is not actually held over await point, remove suppression once
+    //  fixed upstream
+    #[allow(clippy::await_holding_lock)]
+    async fn derive_next_seed(
+        &self,
+        seed: PotSeed,
+        slot_iterations: NonZeroU32,
+    ) -> Option<PotSeed> {
+        let cache_key = CacheKey {
+            seed,
+            slot_iterations,
+        };
+
+        loop {
+            let mut cache = self.cache.lock();
+            let maybe_cache_value = cache.peek(&cache_key).cloned();
+            if let Some(cache_value) = maybe_cache_value {
+                drop(cache);
+                let correct_checkpoints = cache_value.checkpoints.lock().await;
+                if let Some(correct_checkpoints) = correct_checkpoints.as_ref() {
+                    return Some(correct_checkpoints.output().seed());
+                }
+
+                // There was another verification for these inputs and it wasn't successful,
+                // retry
+                continue;
+            }
+
+            let cache_value = CacheValue {
+                checkpoints: Arc::default(),
+            };
+            let checkpoints = Arc::clone(&cache_value.checkpoints);
+            let mut checkpoints = checkpoints
+                .try_lock()
+                .expect("No one can access this mutex yet; qed");
+            // Store pending verification entry in cache
+            cache.push(cache_key, cache_value);
+            // Take a lock before anyone else
```

Comment:
this comment is not required, no ?

Reply:
It needs to be moved up a few lines, I reordered code, but forgot to move the comment accordingly.